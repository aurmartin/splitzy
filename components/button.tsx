import { Pressable, type PressableProps } from "@/components/pressable";
import { Text } from "@/components/text";
import { Colors } from "@/lib/constants";
import { StyleProp, StyleSheet, ViewStyle } from "react-native";

interface ButtonProps extends PressableProps {
  children: React.ReactNode;
  style?: StyleProp<ViewStyle>;
  type?: "primary" | "secondary";
}

export default function Button(props: ButtonProps) {
  const { type = "primary", children, style, ...rest } = props;

  return (
    <Pressable
      {...rest}
      style={[styles.button, styles[type], style]}
      android_ripple={{ color: Colors.primaryLight }}
    >
      <Text style={styles.text}>{children}</Text>
    </Pressable>
  );
}

const styles = StyleSheet.create({
  primary: {
    backgroundColor: Colors.primary,
  },
  secondary: {
    backgroundColor: Colors.secondary,
  },
  button: {
    borderRadius: 40,
    height: 40,
    alignItems: "center",
    justifyContent: "center",
    paddingHorizontal: 16,
  },
  text: {
    color: "white",
  },
});
