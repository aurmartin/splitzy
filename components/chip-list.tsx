import { primaryHue } from "@/lib/constants";
import React from "react";
import { Pressable, StyleProp, View, ViewStyle } from "react-native";
import { Text } from "./text";

interface ChipProps {
  label: string;
  isSelected: boolean;
  onPress: () => void;
}

function Chip(props: ChipProps) {
  const { label, isSelected, onPress } = props;

  const backgroundColor = React.useMemo(() => {
    if (isSelected) {
      return `hsl(${primaryHue}, 50%, 80%)`;
    }

    return "hsl(0, 0%, 90%)";
  }, [isSelected]);

  const style = {
    padding: 8,
    backgroundColor,
    borderRadius: 16,
  };

  return (
    <Pressable
      onPress={onPress}
      style={style}
      accessibilityLabel={label}
      accessibilityRole="button"
    >
      <Text type="bodyMedium">{label}</Text>
    </Pressable>
  );
}

interface ChipListItem {
  label: string;
  value: string;
}

interface ChipListProps {
  onChange: (value: string) => void;
  value: string | string[];
  items: ChipListItem[];
  style?: StyleProp<ViewStyle>;
}

function ChipList(props: ChipListProps) {
  const { onChange, value, items, style } = props;

  const isSelected = (item: ChipListItem) => {
    if (Array.isArray(value)) {
      return value.includes(item.value);
    }
    return item.value === value;
  };

  return (
    <View style={[{ flexDirection: "row", gap: 8 }, style]}>
      {items.map((item) => (
        <Chip
          key={item.value}
          label={item.label}
          isSelected={isSelected(item)}
          onPress={() => onChange(item.value)}
        />
      ))}
    </View>
  );
}

export { ChipList, type ChipListProps };
