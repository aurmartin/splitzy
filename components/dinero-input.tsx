import dinero, { Dinero } from "dinero.js";
import React from "react";
import FloatInput from "./float-input";
import { getMinorUnitAmount } from "@/lib/currency";

const DineroInput = (props: {
  value: Dinero;
  onChange: (value: Dinero) => void;
}) => {
  const { value, onChange: onChangeProp } = props;

  const onChange = React.useCallback(
    (numberValue: number) => {
      const currency = value.getCurrency();

      onChangeProp(
        dinero({
          amount: getMinorUnitAmount(numberValue, currency),
          currency,
        }),
      );
    },
    [onChangeProp, value],
  );

  return <FloatInput value={value.toUnit()} onChange={onChange} />;
};

export default DineroInput;
