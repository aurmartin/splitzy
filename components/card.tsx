import { Pressable } from "@/components/pressable";
import { View } from "react-native";

interface CardProps {
  children: React.ReactNode;
  onPress?: () => void;
}

export default function Card(props: CardProps) {
  if (props.onPress) {
    return (
      <Pressable
        style={{
          borderRadius: 8,
          backgroundColor: "white",
          overflow: "hidden",
          padding: 14,
        }}
        android_ripple={{ color: "hsl(0, 0, 40%)" }}
        onPress={props.onPress}
      >
        {props.children}
      </Pressable>
    );
  } else {
    return (
      <View
        style={{
          backgroundColor: "white",
          borderRadius: 8,
          padding: 14,
        }}
      >
        {props.children}
      </View>
    );
  }
}
