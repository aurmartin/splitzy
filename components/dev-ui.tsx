import { useSystem } from "@/components/system-provider";
import { ExpensesRepository } from "@/lib/expenses/repository";
import { getGroupRow, useAddGroup } from "@/lib/groups";
import { System } from "@/lib/system";
import { generateId } from "@/lib/utils";
import dinero from "dinero.js";
import { usePathname } from "expo-router";
import { useEffect, useState } from "react";
import { Button, Platform, StyleSheet, Text, View } from "react-native";
import { type PercentageSplit } from "@/lib/expenses";

function getRandomMember(members: string[]) {
  return members[Math.floor(Math.random() * members.length)];
}

async function generateExpenses(system: System, groupId: string) {
  const count = 10;
  const group = getGroupRow(system, groupId);
  if (!group) {
    throw new Error("Group not found");
  }
  const members = group.members;

  for (let i = 0; i < count; i++) {
    const payer = getRandomMember(members);

    const total = dinero({
      amount: Math.floor(Math.random() * 100),
      currency: group.currency,
    });

    const splitExpense: PercentageSplit = {
      type: "percentage",
      total,
      members,
      ratios: members.reduce(
        (acc, member) => {
          acc[member] = Math.floor(Math.random() * 100);
          return acc;
        },
        {} as PercentageSplit["ratios"],
      ),
    };

    const id = generateId();
    const date = new Date();
    date.setDate(date.getDate() - i * 3);

    await ExpensesRepository.insertExpense(system, {
      id,
      groupId,
      title: `Expense ${i + 1}`,
      payerName: payer,
      splitExpense,
      createdAt: date,
      updatedAt: date,
      deletedAt: undefined,
      receipt: null,
    });
  }
}

const MemoryUsageInfo = () => {
  const [used, setUsed] = useState(0);

  const formatBytes = (bytes: number) => {
    if (bytes < 1024) {
      return `${bytes} bytes`;
    }
    if (bytes < 1024 * 1024) {
      return `${(bytes / 1024).toFixed(2)} Kb`;
    }
    return `${(bytes / 1024 / 1024).toFixed(2)} Mb`;
  };

  useEffect(() => {
    if (Platform.OS === "web") {
      const interval = setInterval(() => {
        setUsed((performance as any).memory.usedJSHeapSize);
      }, 500);
      return () => clearInterval(interval);
    }
  }, []);

  return <Text style={{ marginBottom: 10 }}>Memory: {formatBytes(used)}</Text>;
};

export default function DevUi() {
  const pathname = usePathname();
  const isGroup = pathname.includes("/groups/");
  const groupId = isGroup ? pathname.split("/groups/")[1] : undefined;
  const system = useSystem();
  const addGroup = useAddGroup();
  const onGenerateExpenses = () => {
    if (!isGroup) {
      return;
    }
    generateExpenses(system, groupId!);
  };

  const onGenerateGroup = () => {
    addGroup({
      name: "Test Group",
      members: ["Alice", "Bob", "Charlie"],
      currency: "EUR",
    });
  };

  return (
    <View style={styles.wrapper}>
      <View style={styles.container}>
        <Text>Dev UI</Text>
        <MemoryUsageInfo />
        <Button
          title="Generate expenses"
          onPress={onGenerateExpenses}
          disabled={!isGroup}
        />
        <Button title="Generate group" onPress={onGenerateGroup} />
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  wrapper: {
    position: "absolute",
    top: 0,
    right: 0,
  },
  container: {
    backgroundColor: "white",
    borderWidth: 1,
    borderColor: "#ccc",
    borderRadius: 4,
    margin: 10,
    padding: 10,
    shadowColor: "black",
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.25,
    shadowRadius: 10,
    gap: 10,
  },
});
