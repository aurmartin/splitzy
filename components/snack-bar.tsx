import React from "react";
import { View } from "react-native";
import { Text } from "@/components/text";
import { SafeAreaView } from "react-native-safe-area-context";

type SnackBarSpec = {
  message: string;
  type: "success" | "error";
};

type SnackBarContextState = {
  show: (message: string, type: "success" | "error") => void;
};

const SnackBar = (props: { snack: SnackBarSpec }) => {
  const { snack } = props;

  const backgroundColor = React.useMemo(() => {
    switch (snack.type) {
      case "success":
        return "hsl(120, 100%, 90%)";
      case "error":
        return "hsl(0, 100%, 90%)";
      default:
        throw new Error("Invalid snack type");
    }
  }, [snack]);

  const color = React.useMemo(() => {
    switch (snack.type) {
      case "success":
        return "hsl(120, 100%, 15%)";
      case "error":
        return "hsl(0, 100%, 15%)";
      default:
        throw new Error("Invalid snack type");
    }
  }, [snack]);

  return (
    <SafeAreaView
      style={{ flex: 1, position: "absolute", bottom: 0, left: 0, right: 0 }}
    >
      <View
        style={{
          backgroundColor,
          margin: 8,
          marginTop: 16,
          padding: 12,
          borderRadius: 4,
          elevation: 3,
        }}
      >
        <Text style={{ color }}>{snack.message}</Text>
      </View>
    </SafeAreaView>
  );
};

const SnackBarContext = React.createContext<SnackBarContextState | null>(null);

const SnackBarProvider = (props: { children: React.ReactNode }) => {
  const { children } = props;

  const [snack, setSnack] = React.useState<SnackBarSpec | null>(null);

  const timerRef = React.useRef<NodeJS.Timeout | null>(null);

  React.useEffect(() => {
    return () => {
      if (timerRef.current) {
        clearTimeout(timerRef.current);
      }
    };
  }, []);

  const show = React.useCallback(
    (message: string, type: "success" | "error") => {
      setSnack({ message, type });

      if (timerRef.current) {
        clearTimeout(timerRef.current);
      }

      timerRef.current = setTimeout(() => setSnack(null), 3000);
    },
    [],
  );

  const value = React.useMemo(() => ({ show }), [show]);

  return (
    <SnackBarContext.Provider value={value}>
      {children}

      {snack && <SnackBar snack={snack} />}
    </SnackBarContext.Provider>
  );
};

const useSnackBar = () => {
  const context = React.useContext(SnackBarContext);

  if (!context) {
    throw new Error("useSnackBar must be used within a SnackBarProvider");
  }

  return context;
};

export { SnackBarProvider, useSnackBar };
