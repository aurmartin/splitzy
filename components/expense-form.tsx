import Button from "@/components/button";
import DineroInput from "@/components/dinero-input";
import Label from "@/components/label";
import Picker from "@/components/picker";
import {
  Split,
  SplitInput,
  changeTotal,
  changeType,
  createReceiptSplit,
  createSplit,
} from "@/components/split-input";
import { Text } from "@/components/text";
import TextInput from "@/components/text-input";
import { useSystem } from "@/components/system-provider";
import { type Expense, type Receipt } from "@/lib/expenses";
import { Group, useMe } from "@/lib/groups";
import { Picker as RNPicker } from "@react-native-picker/picker";
import dinero, { type Dinero } from "dinero.js";
import * as DocumentPicker from "expo-document-picker";
import * as FileSystem from "expo-file-system";
import React from "react";
import { ActivityIndicator, ScrollView, View } from "react-native";
import { useSnackBar } from "./snack-bar";

const ReceiptLoader = () => {
  return (
    <View
      style={{
        flex: 1,
        alignItems: "center",
        justifyContent: "center",
        minHeight: "100%",
        padding: 16,
      }}
    >
      <View>
        <ActivityIndicator
          size="large"
          color="hsl(220, 10%, 50%)"
          style={{ marginBottom: 16 }}
        />
        <Text>Traitement du reçu en cours...</Text>
      </View>
    </View>
  );
};

type ExpenseFormFields = {
  title: Expense["title"];
  payerName: Expense["payerName"];
  splitExpense: Expense["splitExpense"];
  receipt: Expense["receipt"];
};

const ExpenseForm = (props: {
  group: Group;
  expense?: Expense;
  onSubmit: (fields: ExpenseFormFields) => void;
  validationErrors: Record<string, string>;
}) => {
  const { group, expense, onSubmit, validationErrors } = props;

  const system = useSystem();
  const me = useMe(group.id);
  const snackBar = useSnackBar();

  const [title, setTitle] = React.useState("Dépense");

  const [payerName, setPayerName] = React.useState(me);

  const [total, setTotal] = React.useState(
    dinero({ amount: 0, currency: group.currency }),
  );
  const [split, setSplit] = React.useState<Split>(
    createSplit("equal", total, group.members),
  );

  const [receipt, setReceipt] = React.useState<Receipt | null>(null);
  const [receiptLoading, setReceiptLoading] = React.useState(false);

  React.useEffect(() => {
    if (!expense) return;

    setTitle(expense.title);
    setPayerName(expense.payerName);
    setTotal(expense.splitExpense.total);
    setSplit(expense.splitExpense);
    setReceipt(expense.receipt);
  }, [group, expense]);

  const handleTypeChange = (type: Split["type"]) => {
    setSplit((split) => changeType(split, type));
  };

  const handleTotalChange = (total: Dinero) => {
    setTotal(total);
    setSplit((split) => changeTotal(split, total));
  };

  const handleSubmit = () => {
    const fields = {
      title,
      payerName,
      splitExpense: split,
      receipt,
    };

    onSubmit(fields);
  };

  const handleReceipt = async () => {
    const result = await DocumentPicker.getDocumentAsync({
      copyToCacheDirectory: true,
      type: "image/*",
    });
    const uri = result.assets?.[0]?.uri;

    if (!uri) return;

    try {
      setReceiptLoading(true);

      const file = await FileSystem.readAsStringAsync(uri, {
        encoding: FileSystem.EncodingType.Base64,
      });

      const receipt = await system.serverConnector.parseReceipt(file);

      setReceiptLoading(false);
      setReceipt(receipt);
      setTitle(receipt.title);
      setSplit(createReceiptSplit(receipt, group.members));
    } catch (error) {
      if (error instanceof Error) {
        console.error(error.stack);
      } else {
        console.error(error);
      }

      snackBar.show(
        "Une erreur est survenue lors du téléchargement du reçu",
        "error",
      );

      setReceiptLoading(false);
    }
  };

  if (receiptLoading) {
    return <ReceiptLoader />;
  }

  return (
    <ScrollView>
      <View style={{ flex: 1, gap: 16 }}>
        <Button onPress={handleReceipt}>Télécharger un reçu</Button>

        <View>
          <TextInput
            label="Titre"
            placeholder="Titre"
            value={title}
            onChangeText={setTitle}
          />
          {validationErrors.title && (
            <Text style={{ color: "hsl(0, 100%, 30%)" }}>
              {validationErrors.title}
            </Text>
          )}
        </View>

        <View>
          <Label>Montant (€)</Label>
          <DineroInput value={total} onChange={handleTotalChange} />
        </View>

        <View>
          <Picker
            label="Payé par"
            selectedValue={payerName}
            onValueChange={setPayerName}
          >
            {group.members.map((member) => (
              <RNPicker.Item key={member} label={member} value={member} />
            ))}
          </Picker>
          {validationErrors.payerName && (
            <Text style={{ color: "hsl(0, 100%, 30%)" }}>
              {validationErrors.payerName}
            </Text>
          )}
        </View>

        <SplitInput
          value={split}
          onChange={setSplit}
          onTypeChange={handleTypeChange}
        />

        <Button onPress={handleSubmit}>
          {expense ? "Modifier la dépense" : "Créer la dépense"}
        </Button>
      </View>
    </ScrollView>
  );
};

export { ExpenseForm, type ExpenseFormFields };
