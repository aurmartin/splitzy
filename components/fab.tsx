import React, { PropsWithChildren } from "react";
import { StyleSheet, View } from "react-native";
import { Pressable } from "@/components/pressable";
import { Colors } from "../lib/constants";
import { useOptionalSnackBar } from "./snack-bar";

type Props = PropsWithChildren<{
  onPress: () => void;
}>;

export default function FAB(props: Props) {
  const { children, onPress } = props;
  const snackBar = useOptionalSnackBar();

  const bottom = snackBar?.snack ? 20 + 40 : 20;

  return (
    <View style={[styles.wrapper, { bottom }]}>
      <Pressable
        testID="fab-button"
        style={styles.fab}
        onPress={onPress}
        android_ripple={{ color: Colors.primaryLight }}
      >
        {children}
      </Pressable>
    </View>
  );
}

const styles = StyleSheet.create({
  wrapper: {
    position: "absolute",
    bottom: 20,
    right: 8,
    zIndex: 100,
  },
  fab: {
    backgroundColor: Colors.primary,
    padding: 16,
    borderRadius: 16,
    elevation: 2,
    shadowColor: "black",
  },
});
