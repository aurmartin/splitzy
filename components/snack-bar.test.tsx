import { useSnackBar } from "@/components/snack-bar";
import { render, waitFor } from "@/lib/test-utils";
import { render as baseRender } from "@testing-library/react-native";
import { View } from "react-native";
import * as React from "react";

const didRender = jest.fn();

beforeEach(() => {
  jest.clearAllMocks();
});

const TestComponent = (props: {
  message: string;
  type: "success" | "error";
}) => {
  const snack = useSnackBar();
  React.useEffect(
    () => snack.show(props.message, props.type),
    [snack, props.message, props.type],
  );
  didRender();
  return <View />;
};

describe("useSnackBar", () => {
  it("show snackbar success", async () => {
    const tree = render(
      <TestComponent message="test message" type="success" />,
      null,
    );

    await waitFor(() => expect(tree.getByText("test message")));
    expect(didRender).toHaveBeenCalledTimes(1);
    expect(tree).toMatchSnapshot();
  });

  it("show snackbar error", async () => {
    const tree = render(
      <TestComponent message="test message" type="error" />,
      null,
    );

    await waitFor(() => expect(tree.getByText("test message")));
    expect(didRender).toHaveBeenCalledTimes(1);
    expect(tree).toMatchSnapshot();
  });

  it("should throw with invalid type", () => {
    expect(() =>
      render(
        //@ts-ignore
        <TestComponent message="test message" type="invalid" />,
        null,
      ),
    ).toThrow();
  });

  it("should throw without context", () => {
    expect(() =>
      baseRender(<TestComponent message="test message" type="success" />),
    ).toThrow();
  });
});
