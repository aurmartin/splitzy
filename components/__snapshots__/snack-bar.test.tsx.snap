// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useSnackBar show snackbar error 1`] = `
[
  <View />,
  <RNCSafeAreaView
    edges={
      {
        "bottom": "additive",
        "left": "additive",
        "right": "additive",
        "top": "additive",
      }
    }
    style={
      {
        "bottom": 0,
        "flex": 1,
        "left": 0,
        "position": "absolute",
        "right": 0,
      }
    }
  >
    <View
      style={
        {
          "backgroundColor": "hsl(0, 100%, 90%)",
          "borderRadius": 4,
          "elevation": 3,
          "margin": 8,
          "marginTop": 16,
          "padding": 12,
        }
      }
    >
      <Text
        style={
          [
            {
              "fontSize": 16,
              "fontWeight": "400",
              "lineHeight": 24,
            },
            {
              "color": "hsl(0, 100%, 15%)",
            },
          ]
        }
      >
        test message
      </Text>
    </View>
  </RNCSafeAreaView>,
]
`;

exports[`useSnackBar show snackbar success 1`] = `
[
  <View />,
  <RNCSafeAreaView
    edges={
      {
        "bottom": "additive",
        "left": "additive",
        "right": "additive",
        "top": "additive",
      }
    }
    style={
      {
        "bottom": 0,
        "flex": 1,
        "left": 0,
        "position": "absolute",
        "right": 0,
      }
    }
  >
    <View
      style={
        {
          "backgroundColor": "hsl(120, 100%, 90%)",
          "borderRadius": 4,
          "elevation": 3,
          "margin": 8,
          "marginTop": 16,
          "padding": 12,
        }
      }
    >
      <Text
        style={
          [
            {
              "fontSize": 16,
              "fontWeight": "400",
              "lineHeight": 24,
            },
            {
              "color": "hsl(120, 100%, 15%)",
            },
          ]
        }
      >
        test message
      </Text>
    </View>
  </RNCSafeAreaView>,
]
`;
